deployment test

tgrecojs opened the pull request:
This PR contains the code I'm referring to in my recent comment in #31.

tgrecojs committed:
test(e2e): added working E2E test

tgrecojs committed:
chore(test):

* added AirdropSpace type
* removed unused code in test file

tgrecojs commented on contract/test/tribbles-airdrop/e2e.test.js:
```diff
@@ -0,0 +1,217 @@
+  );
+  // this runCoreEval does not work
+  const name = 'airdrop';
+  const result = await runCoreEval({
```
@dckc this CoreEval test seems to still fail

dckc replied:
symptoms? steps to reproduce?

tgrecojs replied:
the errors stem from noble-hashes. I can't pinpoint exactly why the error is happening, but I've just recorded [this video](https://very-decentralized.myfilebase.com/ipfs/Qmd37r9YSapzz2YQfkVDxc1MvyMRZ7SdqVDTXTk9uDeTkU) where I speak on the issue in a fairly detailed manner.

The steps to reproduce are:

### preliminary steps
```
git fetch <remote-name> tg-deployment-test
cd contract
# check that all deps are installed and everything is working as expected
yarn run test -t
# see that output from test matches test output included in block below
```

#### test output
`yarn run test -t` should end with the following output:
```
1..25
# tests 24
# pass 18
# skip 5
# fail 2
```
If it doesn't, then I suggest running a `yarn install` from the root directory and then retrying. If that doesn't work then let me know and I'll try to figure out a way to get a copy of the proper code over to your machine.

## steps to reproduce error 

```sh
$ cd contract
$ docker-compose up -d
$ yarn run docker:logs
# open up new terminal window
$ make clean start-contract-airdrop
```

The proposal will pass and then the CoreEval will fail, producing the error below:


```sh
agd-1  | 2024-09-02T20:50:18.308Z SwingSet: vat: v1: CORE_EVAL failed: (TypeError#8)
agd-1  | 2024-09-02T20:50:18.309Z SwingSet: vat: v1: TypeError#8: toBytes: no function
agd-1  | 2024-09-02T20:50:18.309Z SwingSet: vat: v1: TypeError: toBytes: no function
agd-1  |  at toBytes (#11989:331)
agd-1  |  at hashC (#11989:343)
agd-1  |  at reduceRight ()
agd-1  |  at map ()
agd-1  |  at generateMerkleRoot (#11989:791)
agd-1  |  at (#11989:981)
agd-1  |  at (#11988:11)
agd-1  |  at apply ()
agd-1  |  at safeEvaluate (/bundled-source/.../node_modules/ses/src/make-safe-evaluator.js:78)
agd-1  |  at (.../vats/src/core/chain-behaviors.js:88)
agd-1  |  at ()
agd-1  |
agd-1  | 2024-09-02T20:50:18.310Z SwingSet: xsnap: v1: TypeError#8 ERROR_NOTE: Sent as error:liveSlots:v1#70006
agd-1  | 2024-09-02T20:50:18.311Z SwingSet: ls: v1: Logging sent error stack (TypeError#8)
agd-1  | 2024-09-02T20:50:18.319Z SwingSet: xsnap: v10: UnhandledPromiseRejectionWarning: (RemoteTypeError(error:liveSlots:v1#70006)#10)
agd-1  | 2024-09-02T20:50:18.319Z SwingSet: xsnap: v10: RemoteTypeError(error:liveSlots:v1#70006)#10: toBytes: no function
agd-1  | 2024-09-02T20:50:18.321Z SwingSet: xsnap: v10: TypeError: toBytes: no function
agd-1  |  at makeError (/bundled-source/.../node_modules/ses/src/error/assert.js:243)
agd-1  |  at decodeErrorCommon (/bundled-source/.../node_modules/@endo/marshal/src/marshal.js:281)
agd-1  |  at decodeFromSmallcaps (/bundled-source/.../node_modules/@endo/marshal/src/encodeToSmallcaps.js:434)
agd-1  |  at fromCapData (/bundled-source/.../node_modules/@endo/marshal/src/marshal.js:356)
agd-1  |  at notifyOnePromise (/bundled-source/.../packages/swingset-liveslots/src/liveslots.js:1225)
agd-1  |  at notify (/bundled-source/.../packages/swingset-liveslots/src/liveslots.js:1249)
agd-1  |  at dispatchToUserspace (/bundled-source/.../packages/swingset-liveslots/src/liveslots.js:1514)
agd-1  |  at runWithoutMetering (/bundled-source/.../packages/swingset-xsnap-supervisor/lib/supervisor-subprocess-xsnap.js:60)
agd-1  |  at ()
```

### brandAuxPublisher failure
this error is shown in the video. it appears before the CoreEval errors, and appears to be separate from the CoreEval failure, but I'm adding the stack trace as its still relevant.

```
agd-1  | 2024-09-02T20:50:01.000Z block-manager: block 6617 begin
agd-1  | 2024-09-02T20:50:01.067Z SwingSet: vat: v1: brandAuxPublisher: new Promise
agd-1  | 2024-09-02T20:50:01.068Z SwingSet: vat: v1: boardAuxAdmin: new Promise
agd-1  | 2024-09-02T20:50:01.072Z SwingSet: vat: v1: cannot save non-durable: brandAuxPublisher Object [Alleged: BrandAuxPublisher] { publishBrandInfo: [Function: publishBrandInfo] }
agd-1  | 2024-09-02T20:50:01.073Z SwingSet: vat: v1: brandAuxPublisher settled; remaining: [ 'anchorKits', 'bankMints', 'boardAuxAdmin', 'contractGovernor', 'reserveGovernorCreatorFacet', 'scaledPriceAuthorityKits', 'vBankKits', 'vaultFactoryGovernorCreator' ]
agd-1  | 2024-09-02T20:50:01.074Z SwingSet: vat: v1: cannot save non-durable: boardAuxAdmin Object [Alleged: BoardAuxAdmin] { publishBrandInfo: [Function: publishBrandInfo], init: [Function: init], update: [Function: update] }
agd-1  | 2024-09-02T20:50:01.075Z SwingSet: vat: v1: boardAuxAdmin settled; remaining: [ 'anchorKits', 'bankMints', 'contractGovernor', 'reserveGovernorCreatorFacet', 'scaledPriceAuthorityKits', 'vBankKits', 'vaultFactoryGovernorCreator' ]
```

dckc replied:
> ...
> The proposal will pass and then the CoreEval will fail, producing the error below:
> 
> ```
> ...
> agd-1  |  at generateMerkleRoot (#11989:791)
> ...
> agd-1  |  at safeEvaluate (/bundled-source/.../node_modules/ses/src/make-safe-evaluator.js:78)
> agd-1  |  at (.../vats/src/core/chain-behaviors.js:88)
> ```

Why is there a call to `generateMerkleRoot` in the CoreEval script? `generateMerkleRoot` takes the large collection of  public keys representing eligible participants as an argument, right? The core-eval script goes on chain. It cannot contain the large collection of keys, right? The merkle root should be generated in advance and provided to the core-eval script as a configuration argument, right?

See also:

 - #72

tgrecojs replied:
yes, updating now..

tgrecojs replied:
@dckc this has been addressed in https://github.com/hindley-milner-systems/dapp-ertp-airdrop/pull/99/commits/fcb20318a75bb1be4c5c52e2a20fd57c18c34b88.

the test code has been updated accordingly.

tgrecojs committed:
feat(pause): added method for pausing contract

tgrecojs committed:
chore(cleanup): removed logs, removed unused argument passed into pauseContract method

tgrecojs committed:
feat(contract): added code and test for contract pause functionality

tgrecojs committed:
test: updated simulateClaim utlity function to take into account errors being thrown within contract

tgrecojs committed:
test: updated simulateClaim utlity function to take into account errors being thrown within contract

tgrecojs commented on contract/src/airdrop/airdropKitCreator.js:
```diff
@@ -321,7 +315,7 @@ export const start = async (zcf, privateArgs, baggage) => {
       creator: {
         pauseContract() {
-          // TODO setOfferFilter
+          zcf.setOfferFilter([messagesObject.makeClaimInvitationDescription()]);
```
Unless i'm missing something, this should close #86

tgrecojs commented on contract/test/tribbles-airdrop/contract.test.js:
```diff
@@ -214,6 +214,51 @@ test('Airdrop ::: happy paths', async t => {
+
+  await E(instance.creatorFacet).pauseContract();
+
+  await t.throwsAsync(
```
this is the test code for #86

tgrecojs committed:
feat(ratios): merging in ratio code

tgrecojs committed:
fix(eslint): fixed linting issues in ratios.test.js file

tgrecojs committed:
chore(test-cleanup):

- ported makeClaimOffer args into test/tribbles-airdrop/actors.js file.
- added jsdoc type annotation for simulateClaim args

tgrecojs committed:
fix(noble-hashes): attempting to patch local noble-hashes library to achieve success when deploying contract

tgrecojs committed:
chore(merkle-tree): removed any instances in which merkle tree construction is referenced within the deploy script

dckc commented on contract/test/tribbles-airdrop/e2e.test.js:
```diff
@@ -143,7 +143,10 @@ test.serial('deploy contract with core eval: airdrop / airdrop', async t => {
+        customTerms: {
+          ...makeTerms(),
+          merkleRoot: merkleTreeAPI.generateMerkleRoot(agoricPubkeys),
+        },
```
This is headed in the right direction, but the e2e version of [runCoreEval](https://github.com/hindley-milner-systems/dapp-ertp-airdrop/blob/dev/contract/tools/e2e-tools.js#L369) doesn't take a `config` option.

I suggest you codegen the `merkleRoot` at build time using `coreScriptOptions` following the pattern of `swaparoo` in `dapp-agoric-basics`:

```js
  config1({
    name: 'swaparoo',
    permit: swapPermit,
    coreScriptOptions: {
      swaparooCommittee: {
        voterAddresses: env.SWAP_GOV_ADDR ? { v1: env.SWAP_GOV_ADDR } : {},
      },
    },
  }),
```
https://github.com/Agoric/dapp-agoric-basics/blob/3c28a51c2d6f4b4769364b19025665d5a83dc699/contract/rollup.config.mjs#L79-L87

The tweak goes in here...

https://github.com/hindley-milner-systems/dapp-ertp-airdrop/blob/fcb20318a75bb1be4c5c52e2a20fd57c18c34b88/contract/rollup.config.mjs#L74-L77


The core eval script has to have an `options: ` line that matches `new RegExp(`options: Fail.*`, 'g')` like this too:

```
  config = {
    options: Fail`missing options config`,
  },
```
https://github.com/Agoric/dapp-agoric-basics/blob/3c28a51c2d6f4b4769364b19025665d5a83dc699/contract/src/swaparoo.proposal.js#L150-L155

maybe change from:

```js
  const {
    options: Fail`missing options config`,
    // must be supplied by caller or template-replaced
    bundleID = config.bundleID,
  } = config?.options?.[contractName] ?? {};
```

to:

```js
  const config = {
    // must be supplied by caller or template-replaced
    options: Fail`missing options config`,
  };
  const {
    // must be supplied by caller or template-replaced
    bundleID = Fail`missing bundleID`,
    merkleRoot,
  } = options?.[contractName] ?? {};
```

dckc replied:
The last part of that suggestion wasn't quite right.

We worked it out in office hours, though:

https://github.com/dckc/dapp-ertp-airdrop/tree/dc-config-pattern 4a7fb2a

tgrecojs committed:
test: pushing for office hours

dckc committed:
chore: handle more config possibilities in assert

tgrecojs committed:
feat(deployment):

succesfully installing airdrop contract bundles (when deployed to local agoric instance running via docker)

tgrecojs committed:
chore:

- removed checkBundle from actors.js file
- updated ,tx.json with latest txn contents

dckc commented on contract/,tx.json:
```diff
@@ -0,0 +1 @@
+{"height":"99414","txhash":"9BB60F9A543D16F37DD2BD06E282D4C54EA7063E16DDA4CC1426B6AED41F1884","codespace":"","code":0,"data":"12260A242F636F736D6F732E62616E6B2E763162657461312E4D736753656E64526573706F6E7365","raw_log":"[{\"msg_index\":0,\"events\":[{\"type\":\"coin_received\",\"attributes\":[{\"key\":\"receiver\",\"value\":\"agoric1rwwley550k9mmk6uq6mm6z4udrg8kyuyvfszjk\"},{\"key\":\"amount\",\"v ...
```
darn it; that `,tx.json` thing is a nuisance. it should go in `.gitignore` at least. We fixed that in some dapps.

tgrecojs replied:
removed and added to .gitignore

dckc commented on contract/src/airdrop.proposal.js:
```diff
@@ -36,100 +37,111 @@ harden(makeTerms);
 const airdropManifest = {
   [startAirdrop.name]: {
     consume: {
+      bankManager: true,
```
bankManager????

tgrecojs replied:
I thought bankManager was needed in order to add a new issuer to vstorage (in order to add a new asset issuer officially accepted by the agoric blockchain) 🤔

tgrecojs replied:
this is removed in latest commit.

dckc replied:
oh... if you're adding a vbank asset, you _will_ need it.

Gee... that's a bug. those powers should be separate. IOU an sdk issue.

Meanwhile, to explain my hasty comment a bit, note the warning in the [BootstrapPowers docs](https://docs.agoric.com/guides/coreeval/permissions.html#selected-bootstrappowers):
> **Warning: this includes the right to spend assets for any account.**


cc @jovonni

dckc replied:
are you adding a vbank asset? i.e. a cosmos denom, for example for IBC usage? I think we talked about that some, but it seems to be out of the scope of the current contract.

dckc commented on contract/src/airdrop.proposal.js:
```diff
@@ -36,100 +37,111 @@ harden(makeTerms);
-      board: true, // to publish boardAux info for NFT brand
-      chainStorage: true, // to publish boardAux info for NFT brand
+      namesByAddress: true,
+      namesByAddressAdmin: true,
```
namesByAddressAdmin???

tgrecojs replied:
removed in latest commit.

dckc commented on contract/start-airdrop-permit.json:
```diff
@@ -0,0 +1,49 @@
+{
```
any particular reason to check in a duplicate version of the permit?

tgrecojs replied:
nope, I'll remove the files produced via the bundling process.

dckc commented on contract/start-airdrop-plan.json:
```diff
@@ -0,0 +1,17 @@
+{
```
this is a product of `agoric run`; mixing the 2-bundle `agoric run` approach and the 1-bundle rollup approach is asking for trouble.

tgrecojs replied:
these files were removed in latest commit.

dckc commented on contract/start-airdrop.js:
```diff
@@ -0,0 +1,191 @@
+// This is generated by writeCoreProposal; please edit!
```
another product of `agoric run`

tgrecojs replied:
removed in latest commit.

tgrecojs committed:
chore(cleanup):

- added function that combines the process of creating a token mint, and gather its associated brand and issuer properties.
- removed logs from makeStateMachine function
- continued process of removing unnecessary values from baggage

tgrecojs committed:
fix:

- removed files produced via agoric run (src/start-airdrop*)
- removed unused powers from airdrop permit
- added ,tx.json to .gitignore

tgrecojs committed:
fix(proposal):

- replaced namesByAddress which was accidentily removed.
- removed namesByAddressAdmin which was the intended to be removed in last commit

tgrecojs commented:
I've addressed your most recent concerns @dckc. I haven't managed to resolve the errors with testing the CoreEval however.

The tests in question are in [`contract/test/tribblesAirdrop/e2e.test.js`](https://github.com/hindley-milner-systems/dapp-ertp-airdrop/blob/tg-deployment-test/contract/test/tribbles-airdrop/e2e.test.js) with 3 of 5 tests producing failures 🤔.

That said, the deployment process is working properly, so i'm wondering if it would be wise to merge the code from this brand into the repo, and then address the 3 specific test issues in a follow-up PR. 


Ideally, resolving the issues with the test is something that I can kick through the uprights fairly quickly & with minimal changes to the codebase, in which case I'd lean to keep this one. If the scenario more closely resembles the inverse of this (minimal changes, short resolution time), then in that case I'd spring for merging this in.

dckc commented:
> ...i'm wondering if it would be wise to merge...

your call. I'm just advising from the peanut gallery.

Meanwhle, anything that gets ci to green, including at least `yarn test`, will make it a lot easier for me to engage here.
 - #57

tgrecojs committed:
fix: commenting out lint

tgrecojs committed:
test(coreEval-fix): updated deployment test to reference the correct start script

tgrecojs committed:
chore(lint): fixed linting issues

tgrecojs committed:
test: removed unused testcode

tgrecojs commented:
merging in this code after meeting with @otoole-brendan. 

will pick up work on #57 afterwards
